### PR TITLE
✨ Feature: Dependency-diff Visualization in the Scorecard Action (version 0 - part 1)

### LINKED ISSUE #772: Feature: Give links to dependency packages in the Dependency-diff Visualization markdown result
Please see an output example [here](https://github.com/aidenwang9867/my_workflow_test/pull/59#issuecomment-1187919878).

Other than the fields that are already there in the report like the dependency `ChangeType`, `AggregateScore`, `PackageName`, `Version`, we might want to add some **links** for users to refer to.

For example, for those with a valid source repo URI, we can add a markdown hyperlink to the dependency's src repo. @laurentsimon  is also suggesting adding deps.dev pages to these dependencies, but we haven't found a reliable answer on which deps.dev page we should give for each dependency.

There could be other options, we'd love to hear about more new ideas ;-)
@azeemshaikh38 @naveensrinivasan

### PR DESCRIPTION
**Pull request by aidenwang9867**
#### What kind of change does this PR introduce?
Introduce the Dependency-diff Visualization in the Scorecard Action (version 0) feature.

#### What is the new behavior (if this is a feature change)?**
This is a new mode of Scorecard Action that can run on a pull request, analyze the dependency-diffs, surface the Scorecard checks for them and visualize them in the **(1) PR comment** and (2) check run annotations. **This PR only introduces (1), (2) will be given in a separate one.**

- [ ] Tests for the changes have been added (for bug fixes/features)
TODO (#issue number): add e2e tests for this feature to make sure the run won't crash on a pull request.

#### Which issue(s) this PR fixes
Fixes #772 (the dependency package deps.dev page will be given in the PR comment if the page exists)

#### Does this PR introduce a user-facing change?
Yes.

```release-note
The Scorecard Action now can run on a pull request to surface Scorecard check results in the check run annotations and visualize the dependency-diffs in the pull request comment.
```

### CODE REVIEW

**Review by naveensrinivasan — COMMENTED** [review]
Thanks, @aidenwang9867! 

Could you please add unit tests? It adds reliability.

It is critical that we have tests.

**Comment by aidenwang9867** [comment]
> Thanks, @aidenwang9867!
> 
> Could you please add unit tests? It adds reliability.
> 
> It is critical that we have tests.

Of course! Should I change this to a draft PR for now? It might take me some extra effort to have the tests covered since I'm not very familiar with testing for an action.

**Comment by naveensrinivasan** [comment]
> > Thanks, @aidenwang9867!
> > Could you please add unit tests? It adds reliability.
> > It is critical that we have tests.
> 
> Of course! Should I change this to a draft PR for now / add the tests in a follow-up PR? It might take me some extra effort to have the tests covered since I'm not very familiar with testing for an action app.

The tests are standard go tests. Can we add unit tests for this PR? Probably certain things are possible with a Unit test, which makes sense to skip for e2e.

**Comment by aidenwang9867** [comment]
> > > Thanks, @aidenwang9867!
> > > Could you please add unit tests? It adds reliability.
> > > It is critical that we have tests.
> > 
> > 
> > Of course! Should I change this to a draft PR for now / add the tests in a follow-up PR? It might take me some extra effort to have the tests covered since I'm not very familiar with testing for an action app.
> 
> The tests are standard go tests. Can we add unit tests for this PR? Probably certain things are possible with a Unit test, which makes sense to skip for e2e.

np, will do :-)

File: entrypoint/dependencydiff/dependency_diff_test.go [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+	}
+}
+
+func asPointerChangeType(ct pkg.ChangeType) *pkg.ChangeType {
+	return &ct
+}
+
+func asPointerStr(s string) *string {
+	return &s
+}
```
Comment (aidenwang9867) [inline entrypoint/dependencydiff/dependency_diff_test.go:177]: Unit tests added for dependency-diff visualization. Thanks for the suggestion! @naveensrinivasan

**Comment by spencerschrock** [comment]
Closing this PR due to age, and also https://github.com/ossf/scorecard/pull/3125

**Pull request closed without merging**